scipy/integrate/_bvp.py

```diff
@@ -0,0 +1,1013 @@
+                        np.asarray(dbc_dp, dtype))
+
+    if S is None:
+        fun_wrapped = fun_p
+    else:
+        def fun_wrapped(x, y, p):
+            f = fun_p(x, y, p)
+            if x[0] == a:
+                f[:, 0] = np.dot(D, f[:, 0])
+                f[:, 1:] += np.dot(S, y[:, 1:]) / (x[1:] - a)
+            else:
+                f += np.dot(S, y) / (x - a)
+            return f
+
+    if fun_jac is not None:
+        if S is None:
+            fun_jac_wrapped = fun_jac_p
+        else:
+            Sr = S[:, :, np.newaxis]
+
+            def fun_jac_wrapped(x, y, p):
+                df_dy, df_dp = fun_jac_p(x, y, p)
+                if x[0] == a:
+                    df_dy[:, :, 0] = np.dot(D, df_dy[:, :, 0])
+                    df_dy[:, :, 1:] += Sr / (x[1:] - a)
+                else:
+                    df_dy += Sr / (x - a)
+
+                return df_dy, df_dp
+
+    return fun_wrapped, bc_wrapped, fun_jac_wrapped, bc_jac_wrapped
+
+
+def solve_bvp(fun, bc, x, y, p=None, S=None, fun_jac=None, bc_jac=None,
+              tol=1e-3, max_nodes=1000, verbose=0):
+    """Solve a boundary-value problem for a system of ODEs.
+
+    This function numerically solves a first order system of ODEs subject to
+    two-point boundary conditions::
+
+        dy / dx = f(x, y, p) + S * y / (x - a), a <= x <= b
+        bc(y(a), y(b), p) = 0
+
+    Here x is a 1-dimensional independent variable, y(x) is a n-dimensional
+    vector-valued function and p is a k-dimensional vector of unknown
+    parameters which is to be found along with y(x). For the problem to be
+    determined there must be n + k boundary conditions, i.e. bc must be
+    (n + k)-dimensional function.
+
+    The last singular term in the right-hand side of the system is optional.
+    It is defined by a n-by-n matrix S, such that the solution must satisfy
+    S y(a) = 0. This condition will be forced during iterations, so it must not
+    contradict boundary conditions. See [2] for the explanation how this term
+    is handled when solving BVPs numerically.
+
+    Problems can be solved in a complex domain as well. In this case y and p
+    are considered to be complex, and f and bc are assumed to be complex-valued
+    functions, but x stays real. Note that f and bc must be complex
+    differentiable (satisfy Cauchy–Riemann conditions) in this case, otherwise
+    you should write your problem for real and imaginary parts separately.
+    To solve a problem in a complex domain, pass an initial guess for y with a
+    complex data type, see below.
+
+    Parameters
+    ----------
+    fun : callable
+        Right-hand side of the system. The calling signature is
+        ``fun(x, y, p)``, where ``x`` is ndarray of shape (m,) and ``y`` is
+        ndarray of shape (n, m), meaning that ``y[:, i]`` corresponds to
+        ``x[i]``, ``p`` is optional with shape (k,). The return value must
```

Comment:
Saying "p is optional" is a bit ambigous. This could be clearer: "The calling signature is `fun(x, y)` or `fun(x, y, p)` if parameters are present. Here, x is ..."